File: lib/src/main/java/com/androidessence/materialdesignspecs/MaterialPalettes.java
package com.androidessence.materialdesignspecs;

import com.androidessence.materialdesignspeclibrary.R;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Random;

/**
 * Utility functions for pulling various material design color palettes.
 *
 * Created by adammcneilly and mauker on 11/26/15.
 */
@SuppressWarnings("WeakerAccess")
public class MaterialPalettes {
    // Our variables
    private static final String PREFIX = "mds_";
    private static final String SEPARATOR = "_";
    private static final String ACCENT_SIGNIFIER = "_A";

    // Color names
    public static final String RED = "red";
    public static final String PINK = "pink";
    public static final String PURPLE = "purple";
    public static final String DEEP_PURPLE = "deeppurple";
    public static final String INDIGO = "indigo";
    public static final String BLUE = "blue";
    public static final String LIGHT_BLUE = "lightblue";
    public static final String CYAN = "cyan";
    public static final String TEAL = "teal";
    public static final String GREEN = "green";
    public static final String LIGHT_GREEN = "lightgreen";
    public static final String LIME = "lime";
    public static final String YELLOW = "yellow";
    public static final String AMBER = "amber";
    public static final String ORANGE = "orange";
    public static final String DEEP_ORANGE = "deeporange";
    public static final String BROWN = "brown";
    public static final String GREY = "grey";
    public static final String BLUE_GREY = "bluegrey";

    public static final String[] COLORS_WITH_ACCENT_NAMES = new String[] {
            RED,
            PINK,
            PURPLE,
            DEEP_PURPLE,
            INDIGO,
            BLUE,
            LIGHT_BLUE,
            CYAN,
            TEAL,
            GREEN,
            LIGHT_GREEN,
            LIME,
            YELLOW,
            AMBER,
            ORANGE,
            DEEP_ORANGE
    };

    public static final String[] COLORS_WITHOUT_ACCENT_NAMES = new String[] {
            BROWN,
            GREY,
            BLUE_GREY
    };

    public static final String[] ALL_COLOR_NAMES = new String[] {
            RED,
            PINK,
            PURPLE,
            DEEP_PURPLE,
            INDIGO,
            BLUE,
            LIGHT_BLUE,
            CYAN,
            TEAL,
            GREEN,
            LIGHT_GREEN,
            LIME,
            YELLOW,
            AMBER,
            ORANGE,
            DEEP_ORANGE,
            BROWN,
            GREY,
            BLUE_GREY
    };

    // Color levels
    public static final String LEVEL_50 = "50";
    public static final String LEVEL_100 = "100";
    public static final String LEVEL_200 = "200";
    public static final String LEVEL_300 = "300";
    public static final String LEVEL_400 = "400";
    public static final String LEVEL_500 = "500";
    public static final String LEVEL_600 = "600";
    public static final String LEVEL_700 = "700";
    public static final String LEVEL_800 = "800";
    public static final String LEVEL_900 = "900";
    public static final String LEVEL_A100 = "A100";
    public static final String LEVEL_A200 = "A200";
    public static final String LEVEL_A400 = "A400";
    public static final String LEVEL_A700 = "A700";

    public static final String[] NON_ACCENT_COLOR_LEVELS = new String[] {
            LEVEL_50,
            LEVEL_100,
            LEVEL_200,
            LEVEL_300,
            LEVEL_400,
            LEVEL_500,
            LEVEL_600,
            LEVEL_700,
            LEVEL_800,
            LEVEL_900
    };

    public static final String[] ACCENT_COLOR_LEVELS = new String[] {
            LEVEL_A100,
            LEVEL_A200,
            LEVEL_A400,
            LEVEL_A700
    };

    public static final String[] COLOR_LEVELS = {
            LEVEL_50,
            LEVEL_100,
            LEVEL_200,
            LEVEL_300,
            LEVEL_400,
            LEVEL_500,
            LEVEL_600,
            LEVEL_700,
            LEVEL_800,
            LEVEL_900,
            LEVEL_A100,
            LEVEL_A200,
            LEVEL_A400,
            LEVEL_A700
    };

    /**
     * List of colors that have been randomly generated to avoid repeating colors.
     */
    private static List<Integer> randomList;

    /**
     * List that contains all the colors.
     */
    private static List<Integer> allColors;

    /**
     * Number of times we have created a random color. If we hit the limit we must start over.
     */
    private static int randomCount;

    private static final String MESSAGE_BAD_COLOR_NAME = "Invalid color name.";
    private static final String MESSAGE_BAD_COLOR_LEVEL = "Invalid color level.";

    /**
     * Builds a list of color resources for a Material Design color palette of a specific color.
     *
     * NOTE: For Brown, Grey, or Blue Grey this will not return any accent colors.
     * @param colorName The name of the color to build the palette for.
     * @return The list of relevant color resources for the specific color name.
     * @throws IllegalAccessException The resource could not be accessed.
     */
    public static List<Integer> getColorsByName(String colorName) throws IllegalAccessException{
        return getColorsByName(colorName, true); // Default true for accents.
    }

    /**
     * Builds a list of color resources for a Material Design color palette of a specific color.
     *
     * NOTE: For Brown, Grey, or Blue Grey this will not return any accent colors.
     * @param colorName The name of the color to build the palette for.
     * @param getAccents A flag variable that will either fetch the accent level colors or not.
     * @return The list of relevant color resources for the specific color name.
     * @throws IllegalAccessException The resource could not be accessed.
     */
    public static List<Integer> getColorsByName(String colorName, boolean getAccents) throws IllegalAccessException{
        // Verify color name
        if(!Arrays.asList(ALL_COLOR_NAMES).contains(colorName)) {
            throw new IllegalArgumentException(MESSAGE_BAD_COLOR_NAME);
        }

        List<Integer> colorList = new ArrayList<>();

        Field[] fields = R.color.class.getFields();

        // Loop through colors and find those that have this color name and meet our convention.
        for(Field field : fields) {
            String key = field.getName();
            // Find color that starts with this name.
            // If get accents is true, return color. If it's not, avoid adding ones that have
            // accent signifier.
            if(key.startsWith(PREFIX + colorName + SEPARATOR) && (getAccents || !key.startsWith(PREFIX + colorName + ACCENT_SIGNIFIER))) {
                colorList.add(field.getInt(null));
            }
        }

        return colorList;
    }

    /**
     * Builds a list of color resources for the Accents of a given color name.
     * @param colorName The color to retrieve accent shades for.
     * @return The color resources for all accent colors of the given name.
     */
    public static List<Integer> getAccentColorsByName(String colorName) throws IllegalAccessException {
        // Verify input
        if(!Arrays.asList(COLORS_WITH_ACCENT_NAMES).contains(colorName)) {
            throw new IllegalArgumentException(MESSAGE_BAD_COLOR_NAME);
        }

        List<Integer> colorList = new ArrayList<>();

        Field[] fields = R.color.class.getFields();

        // Loop through and get ones that meet our criteria
        for(Field field : fields) {
            String key = field.getName();
            if(key.startsWith(PREFIX + colorName + ACCENT_SIGNIFIER)) {
                colorList.add(field.getInt(null));
            }
        }

        return colorList;
    }

    /**
     * Builds a list of color resources for a Material Design color palette of a specific color level.
     *
     * NOTE: For accent levels, this will not return Brown, Grey, or Blue Grey.
     * @param colorLevel The level of the colors to build the palette with.
     * @return The list of relevant color resources for the specific color level.
     * @throws IllegalAccessException The resource could not be accessed.
     */
    public static List<Integer> getColorsByLevel(String colorLevel) throws IllegalAccessException{
        // Verify color level
        if(!Arrays.asList(COLOR_LEVELS).contains(colorLevel)) {
            throw new IllegalArgumentException(MESSAGE_BAD_COLOR_LEVEL);
        }

        List<Integer> colorList = new ArrayList<>();

        Field[] fields = R.color.class.getFields();

        // Loop through colors and find those that have this color level and meet our convention.
        for(Field field : fields) {
            String key = field.getName();
            if(key.startsWith(PREFIX) && key.endsWith(SEPARATOR + colorLevel)) {
                colorList.add(field.getInt(null));
            }
        }

        return colorList;
    }

    /**
     * Retrieves a random material design color.
     * @return A color resource for one of the material design colors.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColor() throws IllegalAccessException{
        List<Integer> colorList = getAllColors();

        // Get random
        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves a random color from the palettes that is not an accent color.
     * @return A color resource for one of the material design colors.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomNonAccentColor() throws IllegalAccessException {
        List<Integer> colorList = new ArrayList<>();

        // For each color, add its non-accent colors to list
        for(String color : ALL_COLOR_NAMES) {
            colorList.addAll(getColorsByName(color, false));
        }

        // Return a random one
        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves a random accent color from the palettes.
     * @return A color resource for one of the material design accent colors.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomAccentColor() throws IllegalAccessException {
        List<Integer> colorList = new ArrayList<>();

        // For each accent level, add its colors to the list.
        for(String level : ACCENT_COLOR_LEVELS) {
            colorList.addAll(getColorsByLevel(level));
        }

        // Return a random one
        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves a random material design color for a given level.
     * @param colorLevel The color level to get a color for.
     * @return A color resource for a random color of the given level.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColorByLevel(String colorLevel) throws IllegalAccessException {
        // Get all colors for this level
        // Input check is done in this function.
        List<Integer> colorList = getColorsByLevel(colorLevel);

        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves a random material design color for the given shade, including any available accent shades.
     * @param colorName The name of the color to get a shade for.
     * @return A color resource for a randomly generated shade of the specified color.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColorByName(String colorName) throws IllegalAccessException{
        return getRandomColorByName(colorName, true);
    }

    /**
     * Determines a random color that does not repeat until a color from all different palettes has
     * been returned.
     * @return A color resource for a random Material Design color.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColorNonRepeating() throws IllegalAccessException {
        if(randomList == null || randomList.isEmpty() || randomCount >= randomList.size()) {
            initColorRandomizer();
        }

        // Get color name from our random list
        int colorIndex = randomList.get(randomCount++);
        return getAllColors().get(colorIndex);
    }

    /**
     * Retrieves a random material design color for the given shade.
     * @param colorName The name of the color to get a shade for.
     * @param useAccents Whether or not to include accent shades in the random selection.
     * @return A color resource for a randomly generated shade of the specified color.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColorByName(String colorName, boolean useAccents) throws IllegalAccessException{
        // Get all colors for this name.
        // Input check is in this function.
        List<Integer> colorList = getColorsByName(colorName, useAccents);

        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves a random color from a list of accepted values.
     * @param acceptedColorNames The color names that are acceptable for the random color returned.
     * @return A color resource for a random Material Design color.
     * @throws IllegalAccessException If a resource cannot be accessed.
     */
    public static Integer getRandomColorWithLimits(String[] acceptedColorNames) throws IllegalAccessException {
        // First do error checking
        if (!Arrays.asList(ALL_COLOR_NAMES).containsAll(Arrays.asList(acceptedColorNames))) {
            throw new IllegalArgumentException(MESSAGE_BAD_COLOR_NAME);
        }

        // Get all possible colors
        List<Integer> colorList = new ArrayList<>();

        for (String name : acceptedColorNames) {
            colorList.addAll(getColorsByName(name));
        }

        // Return a random color from the list.
        Random random = new Random();
        return colorList.get(random.nextInt(colorList.size()));
    }

    /**
     * Retrieves all material design palette colors.
     * @return A list of color resources for all available material design colors.
     */
    public static List<Integer> getAllColors() throws IllegalAccessException{
        if (allColors == null) {
            allColors = new ArrayList<>();
        }
        if (allColors.size() > 0) {
            return allColors;
        }

        Field[] fields = R.color.class.getFields();

        // Add all colors that start with our prefix
        for(Field field : fields) {
            String key = field.getName();
            if(key.startsWith(PREFIX)) {
                allColors.add(field.getInt(null));
            }
        }

        return allColors;
    }

    /**
     * Initializes the list used to keep track of whether or not a color has been returned yet.
     */
    private static void initColorRandomizer() throws IllegalAccessException {
        if(randomList == null) {
            randomList = new ArrayList<>();
        }

        if(randomList.size() > 0) {
            randomList.clear();
        }

        int size = getAllColors().size();

        for(int i = 0; i < size; i++) {
            randomList.add(i);
        }

        Collections.shuffle(randomList);
        randomCount = 0;
    }

    private static void sortFields(Field[] fields) {
        final AlphaNumComparator anComp = new AlphaNumComparator();

        Arrays.sort(fields, new Comparator<Field>() {
            @Override
            public int compare(Field lhs, Field rhs) {
                String lhss = lhs.getName();
                String rhss = rhs.getName();

                return anComp.compare(lhss,rhss);
            }
        });
    }

    public static Field[] getFields() {
        return R.color.class.getFields();
    }

//    private static List<String> getAllColorNames(boolean sorted) {
//        List<String> colors = Arrays.asList(ALL_COLOR_NAMES);
//        if (sorted) {
//            Arrays.sort(ALL_COLOR_NAMES);
//        }
//        return new ArrayList<>(Arrays.asList(ALL_COLOR_NAMES));
//    }
}

File: lib/src/main/java/com/androidessence/materialdesignspecs/MaterialPalettes.kt
package com.androidessence.materialdesignspecs

import com.androidessence.materialdesignspeclibrary.R
import java.lang.reflect.Field
import java.util.*

/**
 * Utility functions for pulling various material design color palettes.
 *
 * Created by adammcneilly and mauker on 11/26/15.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate")
object MaterialPalettes {
    //region Private Properties & Helpers
    private const val PREFIX = "mds_"
    private const val SEPARATOR = "_"
    private const val ACCENT_SIGNIFIER = "_A"

    private const val MESSAGE_BAD_COLOR_NAME = "Invalid color name."
    private const val MESSAGE_BAD_COLOR_LEVEL = "Invalid color level."

    private fun sortFields(fields: Array<Field>) {
        val anComp = AlphaNumComparator()

        Arrays.sort(fields) { lhs, rhs ->
            val lhss = lhs.name
            val rhss = rhs.name

            anComp.compare(lhss, rhss)
        }
    }

    private val fields: Array<Field>
        get() = R.color::class.java.fields
    //endregion

    //region Color Names & Levels
    const val RED = "red"
    const val PINK = "pink"
    const val PURPLE = "purple"
    const val DEEP_PURPLE = "deeppurple"
    const val INDIGO = "indigo"
    const val BLUE = "blue"
    const val LIGHT_BLUE = "lightblue"
    const val CYAN = "cyan"
    const val TEAL = "teal"
    const val GREEN = "green"
    const val LIGHT_GREEN = "lightgreen"
    const val LIME = "lime"
    const val YELLOW = "yellow"
    const val AMBER = "amber"
    const val ORANGE = "orange"
    const val DEEP_ORANGE = "deeporange"
    const val BROWN = "brown"
    const val GREY = "grey"
    const val BLUE_GREY = "bluegrey"

    const val LEVEL_50 = "50"
    const val LEVEL_100 = "100"
    const val LEVEL_200 = "200"
    const val LEVEL_300 = "300"
    const val LEVEL_400 = "400"
    const val LEVEL_500 = "500"
    const val LEVEL_600 = "600"
    const val LEVEL_700 = "700"
    const val LEVEL_800 = "800"
    const val LEVEL_900 = "900"
    const val LEVEL_A100 = "A100"
    const val LEVEL_A200 = "A200"
    const val LEVEL_A400 = "A400"
    const val LEVEL_A700 = "A700"

    @JvmStatic
    val COLORS_WITH_ACCENT_NAMES = arrayOf(RED, PINK, PURPLE, DEEP_PURPLE, INDIGO, BLUE, LIGHT_BLUE, CYAN, TEAL, GREEN, LIGHT_GREEN, LIME, YELLOW, AMBER, ORANGE, DEEP_ORANGE)

    @JvmStatic
    val COLORS_WITHOUT_ACCENT_NAMES = arrayOf(BROWN, GREY, BLUE_GREY)

    @JvmStatic
    val ALL_COLOR_NAMES = arrayOf(RED, PINK, PURPLE, DEEP_PURPLE, INDIGO, BLUE, LIGHT_BLUE, CYAN, TEAL, GREEN, LIGHT_GREEN, LIME, YELLOW, AMBER, ORANGE, DEEP_ORANGE, BROWN, GREY, BLUE_GREY)

    @JvmStatic
    val NON_ACCENT_COLOR_LEVELS = arrayOf(LEVEL_50, LEVEL_100, LEVEL_200, LEVEL_300, LEVEL_400, LEVEL_500, LEVEL_600, LEVEL_700, LEVEL_800, LEVEL_900)

    @JvmStatic
    val ACCENT_COLOR_LEVELS = arrayOf(LEVEL_A100, LEVEL_A200, LEVEL_A400, LEVEL_A700)

    @JvmStatic
    val COLOR_LEVELS = arrayOf(LEVEL_50, LEVEL_100, LEVEL_200, LEVEL_300, LEVEL_400, LEVEL_500, LEVEL_600, LEVEL_700, LEVEL_800, LEVEL_900, LEVEL_A100, LEVEL_A200, LEVEL_A400, LEVEL_A700)
    //endregion

    //region Get Colors By Name & Level
    /**
     * Builds a list of color resources for a Material Design color palette of a specific color. The
     * caller has the ability to exclude accent colors using the [getAccents] parameter.
     *
     * NOTE: For Brown, Grey, or Blue Grey this will not return any accent colors.
     */
    @Throws(IllegalAccessException::class)
    @JvmOverloads
    @JvmStatic
    fun getColorsByName(colorName: String, getAccents: Boolean = true): List<Int> {
        // Verify color name
        if (!ALL_COLOR_NAMES.contains(colorName)) {
            throw IllegalArgumentException(MESSAGE_BAD_COLOR_NAME)
        }

        val colorList = ArrayList<Int>()

        // Loop through colors and find those that have this color name and meet our convention.
        for (field in fields) {
            val key = field.name
            // Find color that starts with this name.
            // If get accents is true, return color. If it's not, avoid adding ones that have
            // accent signifier.
            if (key.startsWith(PREFIX + colorName + SEPARATOR) && (getAccents || !key.startsWith(PREFIX + colorName + ACCENT_SIGNIFIER))) {
                colorList.add(field.getInt(null))
            }
        }

        return colorList
    }

    /**
     * Retrieves a list of colors for multiple color names.
     *
     * TODO: Consider if we can consolidate this and `getColorsByName` somehow.
     * @see getColorsByName
     */
    @Throws(IllegalAccessException::class)
    @JvmOverloads
    @JvmStatic
    fun getColorsByNames(colorNames: Array<String>?, getAccents: Boolean = true): List<Int> {
        if (colorNames == null) {
            throw IllegalArgumentException(MESSAGE_BAD_COLOR_NAME)
        }

        val colorList = ArrayList<Int>()

        for (name in colorNames) {
            colorList.addAll(getColorsByName(name, getAccents))
        }

        return colorList
    }

    /**
     * Builds a list of color resources for the accents of a given color name.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getAccentColorsByName(colorName: String): List<Int> {
        // Verify input
        if (!COLORS_WITH_ACCENT_NAMES.contains(colorName)) {
            throw IllegalArgumentException(MESSAGE_BAD_COLOR_NAME)
        }

        val colorList = ArrayList<Int>()

        // Loop through and get ones that meet our criteria
        for (field in fields) {
            val key = field.name
            if (key.startsWith(PREFIX + colorName + ACCENT_SIGNIFIER)) {
                colorList.add(field.getInt(null))
            }
        }

        return colorList
    }

    /**
     * Builds a list of color resources for a Material Design color palette of a specific color level.
     * NOTE: For accent levels, this will not return Brown, Grey, or Blue Grey.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getColorsByLevel(colorLevel: String): List<Int> {
        // Verify color level
        if (!COLOR_LEVELS.contains(colorLevel)) {
            throw IllegalArgumentException(MESSAGE_BAD_COLOR_LEVEL)
        }

        val colorList = ArrayList<Int>()

        // Loop through colors and find those that have this color level and meet our convention.
        for (field in fields) {
            val key = field.name
            if (key.startsWith(PREFIX) && key.endsWith(SEPARATOR + colorLevel)) {
                colorList.add(field.getInt(null))
            }
        }

        return colorList
    }
    //endregion

    //region Random
    /**
     * List of colors that have been randomly generated to avoid repeating colors.
     */
    private var randomList: MutableList<Int>? = null

    /**
     * Number of times we have created a random color. If we hit the limit we must start over.
     */
    private var randomCount: Int = 0

    /**
     * Retrieves a random material design color from any palette.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomColor(): Int {
        val colorList = getAllColors()
        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Retrieves a random color from the palettes that is not an accent color.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomNonAccentColor(): Int {
        val colorList = ArrayList<Int>()
        for (color in ALL_COLOR_NAMES) {
            colorList.addAll(getColorsByName(color, false))
        }
        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Retrieves a random accent color from the palettes that support accents.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomAccentColor(): Int {
        val colorList = ArrayList<Int>()
        for (level in ACCENT_COLOR_LEVELS) {
            colorList.addAll(getColorsByLevel(level))
        }
        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Retrieves a random material design color for a given level. If this is an accent level,
     * certain colors will be excluded from the result.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomColorByLevel(colorLevel: String): Int? {
        // Get all colors for this level
        // Input check is done in this function.
        val colorList = getColorsByLevel(colorLevel)

        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Determines a random color that does not repeat until a color from all different palettes has
     * been returned.
     *
     * @see randomList
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomColorNonRepeating(): Int {
        if (randomList == null || randomList!!.isEmpty() || randomCount >= (randomList?.size
                        ?: 0)) {
            initColorRandomizer()
        }

        val colorIndex = randomList!![randomCount++]
        return getAllColors()[colorIndex]
    }

    /**
     * Retrieves a random material design color for the given name. The caller has the ability to
     * decide if accent colors can be excluded using the [useAccents] parameter.
     *
     * TODO: We want to add a companion getAccents to [getRandomColorWithLimits] as well,
     * and also rename that to be more consistent with this one.
     */
    @Throws(IllegalAccessException::class)
    @JvmOverloads
    @JvmStatic
    fun getRandomColorByName(colorName: String, useAccents: Boolean = true): Int? {
        // Get all colors for this name.
        // Input check is in this function.
        val colorList = getColorsByName(colorName, useAccents)

        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Retrieves a random color from a list of accepted color names.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getRandomColorWithLimits(acceptedColorNames: Array<String>): Int? {
        // First do error checking
        //TODO: Convert color names and such to lists, to avoid this stuff.
        if (!Arrays.asList(*ALL_COLOR_NAMES).containsAll(Arrays.asList(*acceptedColorNames))) {
            throw IllegalArgumentException(MESSAGE_BAD_COLOR_NAME)
        }

        // Get all possible colors
        val colorList = ArrayList<Int>()

        for (name in acceptedColorNames) {
            colorList.addAll(getColorsByName(name))
        }

        // Return a random color from the list.
        val random = Random()
        return colorList[random.nextInt(colorList.size)]
    }

    /**
     * Initializes the list used to keep track of whether or not a color has been returned yet.
     */
    @Throws(IllegalAccessException::class)
    private fun initColorRandomizer() {
        if (randomList == null) {
            randomList = ArrayList()
        }

        if (randomList?.isNotEmpty() == true) {
            randomList?.clear()
        }

        val size = getAllColors().size

        for (i in 0 until size) {
            randomList?.add(i)
        }

        randomList?.shuffle()
        randomCount = 0
    }
    //endregion

    //region All Colors
    /**
     * List that contains all the colors.
     */
    private var allColors: MutableList<Int>? = null

    /**
     * Retrieves all material design palette colors.
     *
     * @return A list of color resources for all available material design colors.
     */
    @Throws(IllegalAccessException::class)
    @JvmStatic
    fun getAllColors(): List<Int> {
        if (allColors == null) {
            allColors = ArrayList()
        }

        if (allColors?.isNotEmpty() == true) {
            return allColors as MutableList<Int>
        }

        // Add all colors that start with our prefix
        for (field in fields) {
            val key = field.name
            if (key.startsWith(PREFIX)) {
                allColors?.add(field.getInt(null))
            }
        }

        return allColors as MutableList<Int>
    }
    //endregion
}
